File: monty_hall_simulation.py
from random import randrange
import sys


# creates the initialize array where each element is one instance of the monty hall problem
# input: num_trials (int)  is the number of experiments to conduct
# output: trials (array of array of ints) is an array where each element is an array with 3 ints
# 0 represents a goat, 1 represents the car
def get_trials_arr(num_trials):
	trials = []

	for i in range(num_trials):

		x = [0, 0, 0]
		index = randrange(3)

		# set one of elements to equal 1
		x[index] = 1

		trials.append(x)

	return trials


# returns the percentage of times you win the car if you stayed
# input: monty_hall_arr is an array representing n Monty Hall problems
# output: float representing the number of times you succeed by sticking to the initial choice
def percentage_success_stay_strategry(monty_hall_arr):
	num_successes = 0

	for i in range(len(monty_hall_arr)):

		chosen_index = randrange(3)

		if monty_hall_arr[i][chosen_index] == 1:
			num_successes += 1

	return (num_successes/len(monty_hall_arr) * 100.0)


# returns the percentage of time you win the car if you changed your choice
def percentage_success_change_strategy(monty_hall_arr):
	num_successes = 0

	for i in range(len(monty_hall_arr)):

		trial = monty_hall_arr[i]

		indices = [0, 1, 2]

		chosen_index = randrange(3)

		indices.remove(chosen_index)

		index = get_closed_room_index(trial, indices)

		if trial[index] == 1:
			num_successes += 1

	return (num_successes/len(monty_hall_arr)* 100.0)


# returns an index of the room that was not reveal
def get_closed_room_index(trial, other_rooms):

	index = other_rooms[0]

	if trial[index] == 1:
		return index
	else:
		return other_rooms[1]



args = sys.argv

if len(args) != 2:
	print("Usage: python monty_hall_simulation.py <num_trials>")
	sys.exit(0)

else:
	num_trials = int(args[1])

arr = get_trials_arr(num_trials)

print("Using stay strategy: " + str(percentage_success_stay_strategry(arr)))
print("Using change strategy: " + str(percentage_success_change_strategy(arr)))

File: monty_hall_simulation.c

// code to run the monty hall simulation in c

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <assert.h>
#define NUM_ROOMS 3
#define MAX_ARRAY_SIZE 100000000


int** get_initial_array(int numTrials);
float calculateWithStayStrategy(int** montyHallArray, int size);
float calculateRateChangeChoice(int** montyHallArray, int size);


int main(){

	long int numTrials;

	// setting the seed of the rand function
	srand(time(NULL));

	printf("Please enter the number of experiments to run:\n");
	scanf("%ld", &numTrials);
	assert((numTrials > 0) && (numTrials <= MAX_ARRAY_SIZE));

	int** trials = get_initial_array(numTrials);

	float staySuccessRate = calculateWithStayStrategy(trials, numTrials);
	float changeChoiceSuccessRate = calculateRateChangeChoice(trials, numTrials);

	// deleting trials
	free(trials);

	printf("Number of trials: %ld\n", numTrials);
	printf("Success rate with STAY strategy: %f\n", staySuccessRate);
	printf("Success rate with CHANGE strategy: %f\n", changeChoiceSuccessRate);


}


int** get_initial_array(int numTrials){

	int** trials = (int**) malloc(numTrials * sizeof(int*));

	for(int i = 0; i < numTrials; i++){

		trials[i] = (int *) malloc(NUM_ROOMS * sizeof(int));

		for(int j = 0; j < NUM_ROOMS; j++){
			trials[i][j] = 0;
		}

		// set random 'room' to 1
		int randomIndex = rand() % NUM_ROOMS;
		assert((randomIndex >= 0) && (randomIndex <= 2));

		trials[i][randomIndex] = 1;
	}

	return trials;
}


float calculateWithStayStrategy(int** montyHallArray, int numTrials){

	int i;
	int numOfWins = 0;

	for(i = 0; i < numTrials; i++){

		int chosenRoom = rand() % NUM_ROOMS;
		assert((chosenRoom >= 0) && (chosenRoom <= 2));

		if( montyHallArray[i][chosenRoom] == 1){
			numOfWins++;
		}
	}

	float percentageOfWins = (float) numOfWins/(float)numTrials * 100;

	return percentageOfWins;
}

float calculateRateChangeChoice(int** montyHallArray, int numTrials){

	int i;
	int numOfWins = 0;

	for(i = 0; i < numTrials; i++){

		int chosenRoom = rand() % NUM_ROOMS;
		assert((chosenRoom >= 0) && (chosenRoom <= 2));

		for(int j = 0; j < NUM_ROOMS; j++){

			// skip the room that was initially chosen
			if (j == chosenRoom)
				continue;

			// if one of the other rooms is the car then it results in a win
			else if(montyHallArray[i][j] == 1){
				numOfWins++;
				break;
			}
		}
	}

	float percentageOfWins = (float) numOfWins/(float)numTrials * 100;

	return percentageOfWins;
}
